### PR TITLE
DAML Studio Activeness checks

### LINKED ISSUE #13835: Daml studio is not handling activeness checks under rollback nodes correctly.
# Affected Daml version

2.0, 2.1.1, 2.2.0-snapshot.20220504.9851.0.4c8e027d

# Bug description

Daml studio is not handling activeness checks correctly when rollback nodes are involved.

# To reproduce

Open the following in daml studio

```Haskell
module Main where
import Daml.Script
import DA.Exception

template Foo
  with
    owner : Party
  where
    signatory owner
    nonconsuming choice Catch : ()
      controller owner
        do try do
              exercise self Fail
            catch
              GeneralError _ -> pure ()
    nonconsuming choice Fail : ()
      controller owner
        do  exercise self Archive
            abort ""

test: Script ()
test = script do
  a <- allocateParty "a"
  c <- submit a do
    createCmd Foo with
      owner = a
  submit a do
    exerciseCmd c Catch
  submit a do
    exerciseCmd c Catch
```

Click on Script results & Show transaction view.
The script fails at the second `submit`, with the following error & transaction:

```
Script execution failed on commit at Main:29:3:
  Attempt to exercise a consumed contract #0:0 (Main:Foo)
  Consumed by: #1:3

Ledger time: 1970-01-01T00:00:00Z

Partial transaction:

Committed transactions: 
  TX 0 1970-01-01T00:00:00Z (Main:24:8)
  #0:0
  │   consumed by: #1:3
  │   referenced by #1:0, #1:2, #1:3
  │   disclosed to (since): 'a' (0)
  └─> create Main:Foo
      with
        owner = 'a'
  
  TX 1 1970-01-01T00:00:00Z (Main:27:3)
  #1:0
  │   disclosed to (since): 'a' (1)
  └─> 'a' exercises Catch on #0:0 (Main:Foo)
      children:
      #1:1
      │   disclosed to (since): 'a' (1)
      └─> rollback
          children:
          #1:2
          │   disclosed to (since): 'a' (1)
          └─> 'a' exercises Fail on #0:0 (Main:Foo)
              children:
              #1:3
              │   disclosed to (since): 'a' (1)
              └─> 'a' exercises Archive on #0:0 (Main:Foo)
```

# Expected behavior

The script should succeed because the consuming action (the `Archive`) is under a rollback node.

# Additional context

Issue reported here: https://discuss.daml.com/t/odd-behavior-with-rollback-nodes-and-archiving/4500

### PR DESCRIPTION
**Pull request by carlpulley-da**
fixes #13835
    
CHANGELOG_BEGIN
  - [Daml Studio] Fix a bug where some contracts were considered inactive even though their consumption was under a rollback. See https://github.com/digital-asset/daml/issues/13835 for details
CHANGELOG_END

- [x] duplicate key checks (movement)
- [x] active keys
- [x] update node info without tree traversal
- [ ] simply traversal to only use reference by, consume by and rollback fields (TODO: will be tracked/managed via https://github.com/digital-asset/daml/issues/14183)
- [ ] code simplification/refactor (TODO: will be tracked/managed via https://github.com/digital-asset/daml/issues/14183)
- [x] extend protobuf with active contracts update

### CODE REVIEW

File: compiler/scenario-service/protos/scenario_service.proto [hunk 1]
```diff
@@ -428,6 +429,7 @@ message ScenarioResult {
   sfixed64 final_time = 4;
   repeated TraceMessage trace_log = 5;
   repeated WarningMessage warnings = 6;
+  repeated string active_contracts = 7;
```
Comment (cocreature) [inline compiler/scenario-service/protos/scenario_service.proto:432]: from a quick look, i think the issue is not on the Haskell side. We’re just not setting that field anywhere in Scala. We need to set this in `Conversions.scala`.

File: compiler/scenario-service/server/src/main/scala/com/digitalasset/daml/lf/scenario/Conversions.scala [hunk 2]
```diff
@@ -53,6 +53,11 @@ final class Conversions(
       .addAllScenarioSteps(steps.asJava)
       .setReturnValue(convertSValue(svalue))
       .setFinalTime(ledger.currentTime.micros)
+      .addAllActiveContracts(
+        ledger.ledgerData.activeContracts.toList
```
Comment (cocreature) [inline compiler/scenario-service/server/src/main/scala/com/digitalasset/daml/lf/scenario/Conversions.scala:57]:
```suggestion
        ledger.ledgerData.activeContracts.view
```

File: compiler/scenario-service/server/src/main/scala/com/digitalasset/daml/lf/scenario/Conversions.scala [hunk 3]
```diff
@@ -67,6 +72,11 @@ final class Conversions(
       .addAllNodes(nodes.asJava)
       .addAllScenarioSteps(steps.asJava)
       .setLedgerTime(ledger.currentTime.micros)
+      .addAllActiveContracts(
+        ledger.ledgerData.activeContracts.toList
```
Comment (cocreature) [inline compiler/scenario-service/server/src/main/scala/com/digitalasset/daml/lf/scenario/Conversions.scala:76]:
```suggestion
        ledger.ledgerData.activeContracts.view
```

File: compiler/scenario-service/server/src/main/scala/com/digitalasset/daml/lf/scenario/Conversions.scala (hunk 2 above)
Comment (carlpulley-da) [reply compiler/scenario-service/server/src/main/scala/com/digitalasset/daml/lf/scenario/Conversions.scala]: Its possible to just drop the `toList` altogether

File: compiler/scenario-service/server/src/main/scala/com/digitalasset/daml/lf/scenario/Conversions.scala [hunk 4]
```diff
@@ -67,6 +72,11 @@ final class Conversions(
       .addAllNodes(nodes.asJava)
       .addAllScenarioSteps(steps.asJava)
       .setLedgerTime(ledger.currentTime.micros)
+      .addAllActiveContracts(
+        ledger.ledgerData.activeContracts
+          .map[String](coid => coidToEventId(coid).toLedgerString)
```
Comment (cocreature) [inline compiler/scenario-service/server/src/main/scala/com/digitalasset/daml/lf/scenario/Conversions.scala:77]: i think the view is still more efficient so you don’t build up the intermediate data structure.

File: daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/Pretty.scala [hunk 5]
```diff
@@ -21,10 +21,14 @@ private[lf] object Pretty {
...
           text(s"that becomes effective at $effectiveAt")
-      case Error.ContractNotActive(coid, tid, consumedBy) =>
+      case Error.ContractNotActive(coid, tid, Some(consumedBy)) =>
         text("Scenario failed due to a fetch of a consumed contract") & prettyContractId(coid) &
           char('(') + (prettyIdentifier(tid)) + text(").") /
           text("The contract had been consumed in transaction") & prettyEventId(consumedBy)
+      case Error.ContractNotActive(coid, tid, None) =>
+        text("Scenario failed due to a fetch of a consumed contract") & prettyContractId(coid) &
+          char('(') + (prettyIdentifier(tid)) + text(").") /
+          text("The contract had been consumed in transaction")
```
Comment (carlpulley-da) [inline daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/Pretty.scala:31]: For the case where the consumedBy contract ID is non-existent, we probably want to modify the error reporting text here?

File: compiler/scenario-service/server/src/main/scala/com/digitalasset/daml/lf/scenario/Conversions.scala [hunk 6]
```diff
@@ -229,13 +239,19 @@ final class Conversions(
...
             .setContractRef(mkContractRef(coid, tid))
             .setConsumedBy(convertEventId(consumedBy))
             .build
         )
+      case Error.ContractNotActive(coid, tid, None) =>
+        builder.setScenarioContractNotActive(
+          proto.ScenarioError.ContractNotActive.newBuilder
+            .setContractRef(mkContractRef(coid, tid))
+            .build
+        )
```
Comment (cocreature) [inline compiler/scenario-service/server/src/main/scala/com/digitalasset/daml/lf/scenario/Conversions.scala:254]:
```suggestion
      case Error.ContractNotActive(coid, tid, optConsumedBy) =>
        val errorBuilder =   proto.ScenarioError.ContractNotActive.newBuilder
            .setContractRef(mkContractRef(coid, tid))
        optConsumedBy.foreach(consumedBy => builder.setConsumedBy(convertEventId(consumedBy))
        builder.setScenarioContractNotActive(errorBuilder.build)
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/scenario/ScenarioLedger.scala [hunk 7]
```diff
@@ -415,163 +415,156 @@ object ScenarioLedger {
...
-                    transaction = trId,
-                    effectiveAt = richTr.effectiveAt,
-                    disclosures = Map.empty,
-                    referencedBy = Set.empty,
-                    consumedBy = None,
-                    rolledbackBy = mbRollbackAncestorId,
-                    parent = mbParentId.map(EventId(trId.id, _)),
+
+              case nc: Node.Create =>
+                val newCache1 = newCache.createdIn(nc.coid, eventId)
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/scenario/ScenarioLedger.scala:472]: It looks like `markAsActive` is now unused and I believe the same holds for `addKey` and `removeKey`. Can we delete them?

File: daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/Pretty.scala [hunk 8]
```diff
@@ -21,10 +21,14 @@ private[lf] object Pretty {
         text(s"Scenario failed due to a fetch of an inactive contract") & prettyContractId(coid) &
           char('(') + (prettyIdentifier(tid)) + text(").") &
           text(s"that becomes effective at $effectiveAt")
-      case Error.ContractNotActive(coid, tid, consumedBy) =>
+      case Error.ContractNotActive(coid, tid, Some(consumedBy)) =>
         text("Scenario failed due to a fetch of a consumed contract") & prettyContractId(coid) &
           char('(') + (prettyIdentifier(tid)) + text(").") /
           text("The contract had been consumed in transaction") & prettyEventId(consumedBy)
+      case Error.ContractNotActive(coid, tid, None) =>
+        text("Scenario failed due to a fetch of a consumed contract") & prettyContractId(coid) &
```
Comment (cocreature) [inline daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/Pretty.scala:29]:
```suggestion
        text("Scenario failed due to a fetch of an inactive contract") & prettyContractId(coid) &
```

File: daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/Pretty.scala (hunk 5 above)
Comment (cocreature) [inline daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/Pretty.scala:31]:
```suggestion
          text("The create of this contract has been rolled back")
```

**Review by cocreature — APPROVED** [review]
Nice work thank you! A few minor comments left.

Apart from that I think there is only one thing left:

You need to add a changelog entry in your commit message (any commit works, can be a new one).

Doesn’t have to be long. Something like

```
changelog_begin

- [Daml Studio] Fix a bug where some contracts were considered inactive even though their consumption was under a rollback. See https://github.com/digital-asset/daml/issues/13835 for details

changelog_end
```

I usually also recommend to include
```
fixes #13835
```
in commit messages to autoclose the corresponding issue.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/scenario/ScenarioLedger.scala [hunk 9]
```diff
@@ -415,163 +415,156 @@ object ScenarioLedger {
 
     @tailrec
     def processNodes(
-        mbCache0: Either[UniqueKeyViolation, LedgerData],
+        cache0: LedgerData,
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/scenario/ScenarioLedger.scala:418]: for other reviewers:
we’re planning to further cleanup this loop later. But now it has close to no tricky semantic content so this is less crucial (it’s mainly  around updating referencedBy & consumedBy which are display only).

You probably also want to disable this with whitespace changes disabled.

File: daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/Pretty.scala (hunk 5 above)
Comment (cocreature) [reply daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/Pretty.scala:31]: left a comment.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/scenario/ScenarioLedger.scala (hunk 7 above)
Comment (carlpulley-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/scenario/ScenarioLedger.scala:472]: It also transpired that `nodeInfoByCoid` was also unused - hence deleted

File: compiler/damlc/lib/DA/Cli/Damlc/Test.hs [hunk 10]
```diff
@@ -263,11 +263,8 @@ prettyErr lfVersion err = case err of
 prettyResult :: SS.ScenarioResult -> DA.Pretty.Doc Pretty.SyntaxClass
 prettyResult result =
     let nTx = length (SS.scenarioResultScenarioSteps result)
-        isActive node =
-            case SS.nodeNode node of
-                Just SS.NodeNodeCreate{} -> isNothing (SS.nodeConsumedBy node)
-                _ -> False
-        nActive = length $ filter isActive (V.toList (SS.scenarioResultNodes result))
+        activeContracts = S.fromList (V.toList (SS.scenarioResultActiveContracts result))
+        nActive = length $ filter (SS.isActive activeContracts) (V.toList (SS.scenarioResultNodes result))
```
[the remark this answers is not in the document]
Comment (carlpulley-da) [reply compiler/damlc/lib/DA/Cli/Damlc/Test.hs:267]: Referenced comment in https://github.com/digital-asset/daml/issues/14183 to ensure this suggested work is tracked/managed.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/scenario/ScenarioLedger.scala (hunk 9 above)
Comment (carlpulley-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/scenario/ScenarioLedger.scala:404]: Ref: https://github.com/digital-asset/daml/issues/14183

**Pull request merged**